Prevent rendering far outside viewport

jridgewell opened the pull request:
This is incomplete, and meant to spur discussion.

- I've prevented rendering outside 3 viewport heights.
  - Should this be increased to e.g. 5?
- I've implemented inside our BaseElement class so that cheap elements
  can override to render further outside or expensive elements closer to
  the viewport.
  - Should we move this to `Resources`, which will give us a little
    better performance?
  - We could also pass data to `#renderOutsideViewport`, which will give
    us a perf boost.

/cc @dvoytenko

dvoytenko commented on src/base-element.js:
```diff
@@ -264,7 +264,10 @@ export class BaseElement {
    */
   renderOutsideViewport() {
-    return true;
+    const box = this.getIntersectionElementLayoutBox();
```
Ideally I'd rather have this method to be very light and very declarative. Runtime generally would have a better idea on how to to calculate the right window. E.g. when we are scrolling down - the window could be made bigger in the direction of scrolling, etc. It'd be very hard to capture these improvements inside of this method.

jridgewell replied:
Moved this to `Resource`. PTAL.

dvoytenko commented on src/service/resources-impl.js:
```diff
@@ -1723,7 +1723,7 @@ export class Resource {
     }
 
-    if (!this.renderOutsideViewport() && !this.isInViewport()) {
+    if (!this.isInViewport() && !this.renderOutsideViewport()) {
```
We also need to discuss the confusion that will be developing between "can I render" and "can I prerender" very soon with @mkhatib 's work. I'm leaning toward a hard and fast rule for now in one place and get the prerendering story done before tuning this deeper.

mkhatib replied:
ha. I had a TODO to explore having `prerenderOutsideViewport` method similar to renderOutsideViewport. For now I'm using `renderOutsideViewport` for both cases (prerender and layout). But yeah we might want to consider to separate these two.

dvoytenko replied:
I'd rather not have `prerenderOutsideViewport`. Instead, ideally we'd do something like:
```
var r = x.rendersOutsideViewport();
var dist = abs(x.position - viewport);
if (dist <= r) {
  x.layout();
} else if (dist <= r * 2) {
  x.prerender();
} else {
  // do nothing at all.
}
```

dvoytenko replied:
@jridgewell could you please also respond to this. Essentially I'd like to move to the system where decision yields one of three results:
- Layout now
- Prerender now
- Do nothing

I think this method is the one that would give us necessary data to calculate this signal.

jridgewell replied:
> I think this method is the one that would give us necessary data to calculate this signal.

We could easily do that by returning an enum value instead of a boolean.

dvoytenko commented on src/service/resources-impl.js:
```diff
@@ -1685,7 +1685,17 @@ export class Resource {
+    // is within X times the viewport height of the current viewport.
+    const viewportBox = this.viewport_.getRect();
+    const distanceFromViewport = this.layoutBox_.top - viewportBox.bottom;
+    return distanceFromViewport <= renders * viewportBox.height;
```
This definitely looks pretty good. One request - could you please take a look in the code where we evaluate the direction of scrolling? We need to see if it'd be useful here as well. At the very least they shouldn't conflict.

jridgewell replied:
I'm not sure how scroll direction would factor in here.

jridgewell commented:
Tests are passing, mind taking a final look?

dvoytenko commented on src/base-element.js:
```diff
@@ -261,10 +261,13 @@ export class BaseElement {
-   * @return {boolean}
+   * Returning a boolean allows or prevents rendering outside the viewport at
+   * any distance, while returning a number allows rendering only when the
+   * element is within X viewports of the current viewport.
```
So, if the element is in the viewport right now, is the response 0 or 1? Let's document this specific case.

jridgewell replied:
If the element is in the viewport, this method is never called.

dvoytenko replied:
Understand. But logically it's very easy to explain what 0 and 1 mean if we bring the reference point to the simplest case. We can say 0 means you are already in viewport. Or we can say that 1 means you are in the next viewport - whichever, as long as we tie in 0 or 1.

jridgewell replied:
Added.

dvoytenko commented on src/service/resources-impl.js:
```diff
@@ -1721,7 +1729,17 @@ export class Resource {
+    // Numeric interface, element is allowed to render outside viewport when it
+    // is within X times the viewport height of the current viewport.
+    const viewportBox = this.resources_.getViewport().getRect();
+    const distanceFromViewport = this.layoutBox_.top - viewportBox.bottom;
```
I'd expect some negative logic here, no? It looks like this value can be negative a lot. So it should either involve abs or two different formulas for above/below viewport.

jridgewell replied:
This was intentionally modeled after the `amp-ad` `#renderOutsideViewport`: If you're below the element, it should render because you've already passed it. I'm fine changing this to `abs`, though we should probably signal to `Element#renderOutsideViewport` that it's above the user.

dvoytenko replied:
Hmm. This is not how `renderOutsideViewport` used to work. This affects performance directly since the user would have to wait to see for what's in the viewport right now longer. I think we should change to `abs` or a above/below formula. Seems like the latter is better since you'd have to do `this.layoutBox_.bottom - viewportBox.top` instead for the above case.

jridgewell replied:
Added.

dvoytenko commented:
Close to LGTM - just would like to confirm the exact values and formulas.

jridgewell commented:
PTAL: Added branching based on the direction of the element from the current viewport. I also (finally) understood what you mean by scroll direction, so I've added a `scrollPenalty` for elements we're scrolling away from.

dvoytenko commented:
@jridgewell I think this looks great. Let's just add tests for resources.js code and we can merge.

jridgewell commented:
Added a bunch of tests.

jridgewell commented:
PTAL.

dvoytenko commented:
LGTM
